exercises/practice/list-ops/src/lib.rs

```diff
@@ -0,0 +1,69 @@
+/// Yields each item of a and then each item of b
+pub fn append<T, I>(_a: I, _b: I) -> impl Iterator<Item = T>
+where
+    I: Iterator<Item = T>,
+{
+    // this empty iterator silences a compiler complaint that
+    // () doesn't implement Iterator
+    std::iter::empty()
+}
+
+/// Combines all items in all nested iterators inside into one flattened iterator
+pub fn concat<I, NI, T>(_nested_iter: I) -> impl Iterator<Item = T>
+where
+    NI: Iterator<Item = T>,
+    I: Iterator<Item = NI>,
+{
+    // this empty iterator silences a compiler complaint that
+    // () doesn't implement Iterator
+    std::iter::empty()
+}
+
+/// Returns an iterator of all items in iter for which `predicate(item)` is true
+pub fn filter<I, T, F>(_iter: I, _predicate: F) -> impl Iterator<Item = T>
+where
+    I: Iterator<Item = T>,
+    F: Fn(&T) -> bool,
+{
+    // this empty iterator silences a compiler complaint that
+    // () doesn't implement Iterator
+    std::iter::empty()
+}
+
+pub fn length<I: Iterator<Item = T>, T>(_iter: I) -> usize {
+    todo!("return the total number of items within iter")
+}
+
+/// Returns an iterator of the results of applying `function(item)` on all iter items
+pub fn map<I, F, T, U>(_iter: I, _function: F) -> impl Iterator<Item = U>
+where
+    I: Iterator<Item = T>,
+    F: Fn(T) -> U,
+{
+    // this empty iterator silences a compiler complaint that
+    // () doesn't implement Iterator
+    std::iter::empty()
+}
+
+pub fn foldl<I, F, T, U>(mut _iter: I, _initial: U, _function: F) -> U
+where
+    I: Iterator<Item = T>,
+    F: Fn(U, T) -> U,
+{
+    todo!("starting with initial, fold (reduce) each iter item into the accumulator from the left")
+}
+
+pub fn foldr<I, F, T, U>(mut _iter: I, _initial: U, _function: F) -> U
+where
+    I: DoubleEndedIterator<Item = T>,
+    F: Fn(U, T) -> U,
```

Comment:
```suggestion
    F: Fn(U, T) -> U,
```
I think with foldr, the argument order is usually reversed compared with foldl, right?
e.g. https://hoogle.haskell.org/?hoogle=foldr

Reply:
I have never used Haskell, but it's not reversed in problem-specifications. It just loops from right to left:

```json
        {
          "uuid": "8066003b-f2ff-437e-9103-66e6df474844",
          "reimplements": "be396a53-c074-4db3-8dd6-f7ed003cce7c",
          "description": "direction dependent function applied to non-empty list",
          "comments": [
            "Reimplemented to remove ambiguity about the parameters of the    ",
            "function input. Expects / to preserve fractions. Integer division",
            "will not work here, since it would compute 4 / 24 = 1 / 6. Use   ",
            "the original test values (be396a53-c074-4db3-8dd6-f7ed003cce7c)  ",
            "if integer division is expected / required."
          ],
          "property": "foldr",
          "input": {
            "list": [1, 2, 3, 4],
            "initial": 24,
            "function": "(acc, el) -> el / acc"
          },
          "expected": 9
        }
```

Reply:
Gets the right answer without reversing the arg order:
```
>> n = 24.0
=> 24.0
>> n = 4 / n
=> 0.16666666666666666
>> n = 3 / n
=> 18.0
>> n = 2 / n
=> 0.1111111111111111
>> n = 1 / n
=> 9.0
```

Reply:
I just checked how other languages do it. Most have the same order, like problem-specifications. Only Haskell reversed it, presumably to stay true to their standard library 🙂 Rust doesn't have foldr, so not much risk of causing confusion. Let's leave it as is.